### PR TITLE
Use more sensible default Java settings for Jenkins [JENKINS-38802]

### PR DESCRIPTION
**Pull request by svanoort**
See [JENKINS-38802](https://issues.jenkins-ci.org/browse/JENKINS-38802)

- [x] Detect major java version to issue recommendations for GC
- [x] Modularize Java arguments to allow separating GC logging from base settings from G1 GC settings
- [x] Conditional flag to determine if we use G1GC vs. normal GC (based on CPU count)
- [x] Test the java version parsing works for OpenJDK and Oracle JRE 7, 8, 9  
- [x] Smoketest script with Debian, CentOS, SUSE

# Known issues
- Java 9 won't accept the settings... but Jenkins won't start on Java 9 anyway due to Jigsaw :(

### CODE REVIEW

**Comment by svanoort** [comment]
> I would probably advise to NOT tune anything unless you really know what you're doing.

@batmat I've got a blog post pending on this (in the process of getting images uploaded and a little formatting cleanup) which walks through the where and why of this. 

But basically:  this autoselects a more reasonable default for Jenkins, based on real-world data.   It's not going to be perfect for everyone but should cover 90% of people better

**Comment by svanoort** [comment]
> I would probably go to 5 or 10 smaller files. 

That's a point. 5x20 MB sounds about right to me.

**Comment by oleg-nenashev** [comment]
I would +1 anything, which enables Groovy class unloading by default. 

But I agree with @batmat that this huge set of options will rather confuse 90% of Jenkins admins, who are not so friendly with Java. In addition, usage of options increases probability of incompatibilities with non-classic Java builds and Java versions, which people may use.

I would vote for introducing such options only for a managed environment like Docker images, but not for generic installers.

So weak 👎 from me, IMHO the approach needs discussion in the developer mailing list.

**Comment by svanoort** [comment]
> IMHO the approach needs discussion in the developer mailing list.

It was brought up and got almost no replies.  

FWIW the more detailed analysis behind these arguments (with quite a bit of data) is working its way along, it's just a slow process getting everything formatted -- intent here is that most users will get the automatic G1 settings (or the default GC settings if they're running with 2 or fewer CPUs or Java 7), which have been tested to be very strong for Jenkins.  

People that know how to customize can turn off the automatic settings and go to town.

**Comment by svanoort** [comment]
I've got the evidence for the GC settings up here on [The CloudBees Blog](https://www.cloudbees.com/blog/joining-big-leagues-tuning-jenkins-gc-responsiveness-and-stability)

Yes, it's not on the jenkins.io blog -- because I'm leveraging anonymized customer data and that's not appropriate to tie into the main Jenkins -- but the overall Jenkins project gets the benefits of that.

**Comment by daniel-beck** [comment]
@svanoort Any plans to get this integrated? It would be a worthwhile addition IMO.

**Comment by svanoort** [comment]
@daniel-beck Yes, I would like to integrate this -- much more simply since we don't have any plans on Java 9 support now.

**Comment by daniel-beck** [comment]
@svanoort Are you seeking to get this integrated? Is this still labeled work-in-progress deliberately?

File: rpm/build/SOURCES/jenkins.sysconfig.in [hunk 1]
```diff
@@ -38,13 +38,34 @@ JENKINS_USER="@@ARTIFACTNAME@@"
...
 
-## Type: string
-## Default:     "-Djava.awt.headless=true"
+## Type:        string
+## Default:     "-server -XX:+AlwaysPreTouch -XX:+ParallelRefProcEnabled"
+## ServiceRestart: @@ARTIFACTNAME@@
+# Sensible defaults, server is normally default but we ensure it is on
+# AlwaysPreTouch prezeroes allocated heap pages at startup, which improves runtime performance at slight cost to startup time
+# ParallelRefProcEnabled enables use of concurrent threads to process references, which Jenkins uses heavily for caching
+BASE_JAVA_ARGS=' -server -XX:+AlwaysPreTouch -XX:+ParallelRefProcEnabled '
```
Comment (samrocketman) [inline rpm/build/SOURCES/jenkins.sysconfig.in]: This is unnecessarily surrounded by spaces.  You already surround the variable with spaces in `JENKINS_JAVA_OPTIONS`.

In other parts of this PR the same feedback applies.  However, I'll only comment this once.

**Comment by svanoort** [comment]
@daniel-beck It just needs testing to integrate -- basically someone needs to manually run it against both sun and openjdk Java 8 for centos, suse, and debian.  Haven't had time to do the manual testing bits though.  The automated testcases don't cover both JVMs because of differing installs. 

If someone can try it out against those and verify it works fine, I think it's mergeable.

**Comment by oleg-nenashev** [comment]
I'd guess this is stalled

**Comment by basil** [comment]
This has been proposed for close for 2 years with no response.

**Pull request closed without merging**